src/step/assemble_constraints_matrix.hpp

```diff
@@ -1,58 +1,8 @@
 #pragma once
 
-#include <Kokkos_Core.hpp>
-#include <Kokkos_Profiling_ScopedRegion.hpp>
-
-#include "src/constraints/constraints.hpp"
-#include "src/solver/copy_constraints_to_sparse_matrix.hpp"
-#include "src/solver/copy_sparse_values_to_transpose.hpp"
-#include "src/solver/solver.hpp"
-
 namespace openturbine {
-inline void AssembleConstraintsMatrix(Solver& solver, Constraints& constraints) {
-    auto region = Kokkos::Profiling::ScopedRegion("Assemble Constraints Matrix");
-
-    if (constraints.num == 0) {
-        return;
-    }
-
-    {
-        auto const_region = Kokkos::Profiling::ScopedRegion("Constraints Matrix");
-        auto constraint_policy =
-            Kokkos::TeamPolicy<>(static_cast<int>(constraints.num), Kokkos::AUTO());
-
-        Kokkos::parallel_for(
-            "CopyConstraintsToSparseMatrix", constraint_policy,
-            CopyConstraintsToSparseMatrix<Solver::CrsMatrixType>{
-                constraints.row_range, constraints.base_node_col_range,
-                constraints.target_node_col_range, solver.B, constraints.base_gradient_terms,
-                constraints.target_gradient_terms
-            }
-        );
-    }
-
-    {
-        auto trans_region = Kokkos::Profiling::ScopedRegion("Transpose Constraints Matrix");
-        auto B_num_rows = solver.B.numRows();
-        auto constraint_transpose_policy = Kokkos::TeamPolicy<>(B_num_rows, Kokkos::AUTO());
-        auto tmp_row_map = Solver::RowPtrType(
-            Kokkos::view_alloc(Kokkos::WithoutInitializing, "tmp_row_map"),
-            solver.B_t.graph.row_map.extent(0)
-        );
-        Kokkos::deep_copy(tmp_row_map, solver.B_t.graph.row_map);
-        Kokkos::parallel_for(
-            "CopySparseValuesToTranspose", constraint_transpose_policy,
-            CopySparseValuesToTranspose<Solver::CrsMatrixType>{solver.B, tmp_row_map, solver.B_t}
-        );
-        KokkosSparse::sort_crs_matrix(solver.B_t);
-    }
+struct Solver;
+struct Constraints;
```

Comment:
This makes me nervous -- why are we using forward declarations here? I think the risks outweigh the compile-time savings here (e.g. unit testing this -- would there be surprises?). For what it's worth, Google's style guild also discourages it.

https://google.github.io/styleguide/cppguide.html#Forward_Declarations

Reply:
Forward declarations are good for decreasing compile times and binary size, but are also for decoupling our code and ensuring that we are doing a good job of managing our dependencies.  

an example: in our unit tests, we created and manipulated the `Constraints` and `StepParameters` structs.  However, we never actually included them, instead getting their code transitively from `Step.hpp`.  A change in our code layout broke our tests and could break user code.  

Their cautionary tales against the forward declaration seem to pertain to using a forward declaration within an implementation.  I agree that this is bad.  I don't think they apply when using forward declarations when defining your interface, which is common practice.